File: go/cross_language_test.go
package ucotron

import (
	"context"
	"os"
	"testing"
)

// TestCrossLanguage* tests run against a live Ucotron server.
// Set UCOTRON_TEST_SERVER_URL to the server URL (e.g., "http://127.0.0.1:8420").
// If the env var is not set, tests are skipped.

func getTestServerURL(t *testing.T) string {
	t.Helper()
	url := os.Getenv("UCOTRON_TEST_SERVER_URL")
	if url == "" {
		t.Skip("UCOTRON_TEST_SERVER_URL not set — skipping integration test")
	}
	return url
}

func testClient(t *testing.T) *Client {
	t.Helper()
	url := getTestServerURL(t)
	return NewClient(url, nil)
}

func TestCrossLanguageHealth(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	health, err := client.Health(ctx)
	if err != nil {
		t.Fatalf("Health() failed: %v", err)
	}

	if health.Status != "ok" {
		t.Errorf("expected status 'ok', got %q", health.Status)
	}
	if health.Version == "" {
		t.Error("expected non-empty version")
	}
	if health.InstanceID == "" {
		t.Error("expected non-empty instance_id")
	}
	if health.InstanceRole == "" {
		t.Error("expected non-empty instance_role")
	}
	if health.StorageMode == "" {
		t.Error("expected non-empty storage_mode")
	}
}

func TestCrossLanguageMetrics(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	metrics, err := client.Metrics(ctx)
	if err != nil {
		t.Fatalf("Metrics() failed: %v", err)
	}

	if metrics.InstanceID == "" {
		t.Error("expected non-empty instance_id")
	}
	if metrics.UptimeSecs < 0 {
		t.Errorf("expected uptime_secs >= 0, got %d", metrics.UptimeSecs)
	}
}

func TestCrossLanguageAddMemory(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	result, err := client.AddMemory(ctx, "Go SDK test: Mount Everest is the tallest mountain.", &AddMemoryOptions{
		Namespace: "go_test",
	})
	if err != nil {
		t.Fatalf("AddMemory() failed: %v", err)
	}

	if result.ChunkNodeIDs == nil {
		t.Error("expected chunk_node_ids to be present")
	}
	if result.EntityNodeIDs == nil {
		t.Error("expected entity_node_ids to be present")
	}
}

func TestCrossLanguageSearch(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	limit := 5
	result, err := client.Search(ctx, "tallest mountain", &SearchOptions{
		Limit:     &limit,
		Namespace: "go_search",
	})
	if err != nil {
		t.Fatalf("Search() failed: %v", err)
	}

	if result.Results == nil {
		t.Error("expected results array")
	}
	if result.Query == "" {
		t.Error("expected non-empty query echo")
	}
}

func TestCrossLanguageAugment(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	result, err := client.Augment(ctx, "What do you know about artificial intelligence?", &AugmentOptions{
		Namespace: "go_aug",
	})
	if err != nil {
		t.Fatalf("Augment() failed: %v", err)
	}

	if result.Memories == nil {
		t.Error("expected memories array")
	}
	if result.Entities == nil {
		t.Error("expected entities array")
	}
	// context_text is always a string (possibly empty)
}

func TestCrossLanguageLearn(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	result, err := client.Learn(ctx, "The user mentioned they prefer dark mode and use VSCode.", &LearnOptions{
		Namespace: "go_learn",
	})
	if err != nil {
		t.Fatalf("Learn() failed: %v", err)
	}

	if result.MemoriesCreated < 0 {
		t.Errorf("expected memories_created >= 0, got %d", result.MemoriesCreated)
	}
}

func TestCrossLanguageListMemories(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	result, err := client.ListMemories(ctx, nil)
	if err != nil {
		t.Fatalf("ListMemories() failed: %v", err)
	}

	// Should return a slice (possibly empty)
	if result == nil {
		t.Error("expected non-nil result")
	}
}

func TestCrossLanguageListEntities(t *testing.T) {
	client := testClient(t)
	ctx := context.Background()

	result, err := client.ListEntities(ctx, nil)
	if err != nil {
		t.Fatalf("ListEntities() failed: %v", err)
	}

	// Should return a slice (possibly empty)
	if result == nil {
		t.Error("expected non-nil result")
	}
}

File: typescript/__tests__/cross_language.test.ts
/**
 * Cross-language SDK integration tests for TypeScript.
 *
 * These tests run against a live Ucotron server. Set UCOTRON_TEST_SERVER_URL
 * to the server URL (e.g., "http://127.0.0.1:8420"). If the env var is not
 * set, tests are skipped.
 */

import { Ucotron } from "../src/client";

const SERVER_URL = process.env.UCOTRON_TEST_SERVER_URL;

const describeIfServer = SERVER_URL ? describe : describe.skip;

describeIfServer("Cross-language integration tests (TypeScript)", () => {
  let client: Ucotron;

  beforeAll(() => {
    client = new Ucotron(SERVER_URL!);
  });

  test("health check returns ok status", async () => {
    const health = await client.health();
    expect(health.status).toBe("ok");
    expect(health.version).toBeTruthy();
    expect(health.instance_id).toBeTruthy();
    expect(health.instance_role).toBeTruthy();
    expect(health.storage_mode).toBeTruthy();
    expect(health.models).toBeDefined();
  });

  test("metrics returns server stats", async () => {
    const metrics = await client.metrics();
    expect(metrics.instance_id).toBeTruthy();
    expect(typeof metrics.total_requests).toBe("number");
    expect(typeof metrics.uptime_secs).toBe("number");
    expect(metrics.uptime_secs).toBeGreaterThanOrEqual(0);
  });

  test("add_memory creates a memory and returns IDs", async () => {
    const ns = `ts_test_${process.pid}`;
    const result = await client.addMemory(
      "TypeScript SDK test: The Eiffel Tower is in Paris, France.",
      { namespace: ns }
    );

    // Response should have the expected shape
    expect(result).toHaveProperty("chunk_node_ids");
    expect(result).toHaveProperty("entity_node_ids");
    expect(result).toHaveProperty("edges_created");
    expect(result).toHaveProperty("metrics");
    expect(Array.isArray(result.chunk_node_ids)).toBe(true);
  });

  test("search returns results with scoring breakdown", async () => {
    const ns = `ts_search_${process.pid}`;
    const result = await client.search("Eiffel Tower in Paris", {
      limit: 5,
      namespace: ns,
    });

    expect(result).toHaveProperty("results");
    expect(result).toHaveProperty("total");
    expect(result).toHaveProperty("query");
    expect(Array.isArray(result.results)).toBe(true);
    expect(result.query).toBeTruthy();
  });

  test("augment returns context for LLM injection", async () => {
    const ns = `ts_aug_${process.pid}`;
    const result = await client.augment(
      "What do you know about artificial intelligence?",
      { namespace: ns }
    );

    expect(result).toHaveProperty("memories");
    expect(result).toHaveProperty("entities");
    expect(result).toHaveProperty("context_text");
    expect(Array.isArray(result.memories)).toBe(true);
    expect(Array.isArray(result.entities)).toBe(true);
    expect(typeof result.context_text).toBe("string");
  });

  test("learn extracts memories from agent output", async () => {
    const ns = `ts_learn_${process.pid}`;
    const result = await client.learn(
      "The user mentioned they prefer dark mode and use VSCode.",
      { namespace: ns }
    );

    expect(result).toHaveProperty("memories_created");
    expect(result).toHaveProperty("entities_found");
    expect(result).toHaveProperty("conflicts_found");
    expect(typeof result.memories_created).toBe("number");
  });

  test("list_memories returns an array", async () => {
    const result = await client.listMemories();
    expect(Array.isArray(result)).toBe(true);
  });

  test("list_entities returns an array", async () => {
    const result = await client.listEntities();
    expect(Array.isArray(result)).toBe(true);
  });
});
